### PR TITLE
Allow semantics labels to be shorter or longer than raw text

### LINKED ISSUE #36210: Providing a semantic label in TextSpans inside a RichText causes an exception in RenderParagraph
Repro code attached (either set useSemanticsDebug = true, or tap the toggle button after the app has launched).

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
 
void main() => runApp(MyApp());
 
class MyApp extends StatefulWidget {
  @override
  State createState() => _MyAppState();
}
 
class _MyAppState extends State<MyApp> {
  bool useSemanticsDebugger = false;
 
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      showSemanticsDebugger: useSemanticsDebugger,
      theme: ThemeData(primaryColor: Colors.blue),
      home: Scaffold(
        appBar: AppBar(
          title: Text('RichText Semantics Test'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Semantics(
                container: true,
                child: RichText(
                  text: TextSpan(
                    style: Theme.of(context).textTheme.body1,
                    children: [
                      TextSpan(text: 'Some Text ',
                               semanticsLabel: '', // <-- this causes a crash.
                      ),
                      TextSpan(
                        text: 'Clickable',
                        recognizer: TapGestureRecognizer()..onTap = () => { }),
                      ])),
              ),
              RaisedButton(
                child: Text('Toggle semantics debugger'),
                onPressed: () =>
                    setState(() =>
                    useSemanticsDebugger = !useSemanticsDebugger),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

### PR DESCRIPTION
**Pull request by dnfield**
## Description

Refactors the way we assemble semantics nodes for text spans, and allows for text span children that have labels shorter than their text (or a label that is an empty string). This also should fix some other bugs where we generally would let the semantics readout get out of sync with what was actually selected, by avoiding using the `toPlaintext` to calculate what's read.

In the course of refactoring this, I've labeled a few relatively new methods and one class as deprecated - it was more complicated to continue using the semantics of those methods than to just write a new method with the information we need. I'm not entirely clear on whether this consists of a breaking change - I think it might be at least a "soft" break but could use another point of view on that.

I'd like to make what I'm pretty sure is a breaking change to get rid of the `if (recognizer is ConcreteInstance) .. `logic by adding some new abstract method that the recognizers we care about could implement, but that can be saved for a separate occasion.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/36210

## Tests

I added the following tests:

A new test where the label is an empty string.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 1]
```diff
@@ -731,49 +731,75 @@ class RenderParagraph extends RenderBox
...
-  // Holds either [GestureRecognizer] or null (for placeholders) to generate
-  // proper semnatics configurations.
-  final List<dynamic> _inlineSemanticsElements = <dynamic>[];
+  /// Collected during [describeSemanticsConfiguration], used by
+  /// [assembleSemanticsNode] and [_combineSemanticsInfo].
+  List<InlineSpanSemanticsInformation> _semanticsInfo;
+
+  /// Combines _semanticsInfo entries where permissible, determined by
+  /// [InlineSpanSemanticsInformation.requiresOwnNode].
+  Iterable<InlineSpanSemanticsInformation> _combineSemanticsInfo() sync* {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/paragraph.dart]: don't use sync* in non-debug code

Comment (dnfield) [reply packages/flutter/lib/src/rendering/paragraph.dart]: I know you've mentioned this before, but we already have many examples of using this in the framework.  

Is it really better to fully materialize this list and then iterate over it again later?

Comment (dnfield) [reply packages/flutter/lib/src/rendering/paragraph.dart]: (I'm not trying to do anything that requires memoization here)

File: packages/flutter/lib/src/painting/inline_span.dart [hunk 2]
```diff
@@ -35,6 +35,70 @@ class Accumulator {
...
+    if (other is! InlineSpanSemanticsInformation) {
+      return false;
+    }
+    return other.text == text && other.semanticsLabel == semanticsLabel && other.recognizer == recognizer && other.isPlaceholder == isPlaceholder;
+  }
+
+  @override
+  int get hashCode => hashValues(text, semanticsLabel, recognizer, isPlaceholder);
+
+
```
Comment (GaryQian) [inline packages/flutter/lib/src/painting/inline_span.dart]: Extra lines

File: packages/flutter/lib/src/painting/inline_span.dart [hunk 3]
```diff
@@ -175,6 +239,28 @@ abstract class InlineSpan extends DiagnosticableTree {
     return buffer.toString();
   }
 
+  /// Flattens the [InlineSpan] tree to a list of plain text strings and
```
Comment (GaryQian) [inline packages/flutter/lib/src/painting/inline_span.dart]: Did you mean to a `list of InlenSpanSemanticsInformation objects`?

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 4]
```diff
@@ -791,85 +817,47 @@ class RenderParagraph extends RenderBox
...
+          ..label = info.semanticsLabel ?? info.text;
+        if (info.recognizer != null) {
+          if (info.recognizer is TapGestureRecognizer) {
+            final TapGestureRecognizer recognizer = info.recognizer;
+            configuration.onTap = recognizer.onTap;
+          } else if (info.recognizer is LongPressGestureRecognizer) {
+            final LongPressGestureRecognizer recognizer = info.recognizer;
+            configuration.onLongPress = recognizer.onLongPress;
+          } else {
+            assert(false);
```
Comment (GaryQian) [inline packages/flutter/lib/src/rendering/paragraph.dart]: Do we not allow the recognizer to be any of the other types of GestureRecognizers?

Comment (dnfield) [reply packages/flutter/lib/src/rendering/paragraph.dart]: Currently we don't - I think we should fix that, but we should do it with something that will probably be breaking.

File: packages/flutter/lib/src/painting/inline_span.dart (hunk 3 above)
Comment (dnfield) [reply packages/flutter/lib/src/painting/inline_span.dart]: Yup, fixing that.

File: packages/flutter/lib/src/painting/inline_span.dart (hunk 2 above)
Comment (dnfield) [reply packages/flutter/lib/src/painting/inline_span.dart]: Fixing

File: packages/flutter/lib/src/painting/inline_span.dart [hunk 5]
```diff
@@ -175,6 +239,28 @@ abstract class InlineSpan extends DiagnosticableTree {
...
+  ///
+  /// [PlaceholderSpan]s in the tree will be represented with a
+  /// [InlineSpanSemanticsInformation.placeholder] value.
+  List<InlineSpanSemanticsInformation> getSemanticsInformation() {
+    final List<InlineSpanSemanticsInformation> collector = <InlineSpanSemanticsInformation>[];
+    computeSemanticsInformation(collector);
+    return collector;
+  }
+
+  /// Walks the [InlineSpan] tree and accumulates the plain text representation
```
Comment (GaryQian) [inline packages/flutter/lib/src/painting/inline_span.dart]: Here too

Comment (dnfield) [reply packages/flutter/lib/src/painting/inline_span.dart]: Done

File: packages/flutter/lib/src/rendering/paragraph.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/rendering/paragraph.dart]: Refactored to use a list instead.

**Review by GaryQian — APPROVED** [review]
LGTM, Thanks!

**Pull request merged**